src/magic_folder/config.py

```diff
@@ -321,6 +389,229 @@ def connect(self, *a, **kw):
+
+    remaining = leaf_candidates - to_discard
+    if len(remaining) != 1:
+        raise ValueError(
+            "Database state inconsistent when loading local snapshot.  "
+            "Leaf candidates: {!r}".format(
+                remaining,
+            ),
+        )
+    [leaf_identifier] = remaining
+    return leaf_identifier
+
+
+def _get_remote_parents(identifier, parents):
+    """
+    Get the identifiers for remote parents for the given snapshot.
+
+    :param unicode identifier: The identifier of the snapshot about which to
+        retrieve information.
+
+    :param dict[unicode, [(int, bool, unicode)]] parents: Information about
+        the parent/child relationships between the snapshots.  See
+        ``_get_parents`` for details.
+
+    :return [unicode]: Identifiers for the remote parents of the identified
+        snapshot.
+    """
+    return list(
+        parent_identifier
+        for (ignored, only_local, parent_identifier)
+        in parents.get(identifier, [])
+        if not only_local
+    )
+
+
+def _get_local_parents(identifier, parents, construct_parent_snapshot):
+    """
+    Get the identifiers for local parents for the given snapshot.
+
+    :param unicode identifier: The identifier of the snapshot about which to
+        retrieve information.
+
+    :param dict[unicode, [(int, bool, unicode)]] parents: Information about
+        the parent/child relationships between the snapshots.  See
+        ``_get_parents`` for details.
+
+    :return [unicode]: Identifiers for the local parents of the identified
+        snapshot.
+    """
+    return list(
+        construct_parent_snapshot(parent_identifier)
+        for (ignored, only_local, parent_identifier)
+        in parents.get(identifier, [])
+        if only_local
+    )
+
+
+def _construct_local_snapshot(identifier, path, author, content_paths, metadata, parents):
+    """
+    Instantiate a ``LocalSnapshot`` corresponding to the given identifier.
+
+    :param unicode identifier: The identifier of the snapshot to instantiate.
+
+    :param unicode path: The path associated with the snapshot.
+
+    :param LocalAuthor author: The author associated with the snapshot.
+
+    :param dict[unicode, unicode] content_paths: A mapping from snapshot
+        identifiers to the filesystem location of the content for that
+        snapshot.
+
+    :param metadata: See ``_get_metadata``.
+
+    :param parents: See ``_get_parents``.
+
+    :return LocalSnapshot: The requested snapshot, populated with information
+        from the given parameters, including fully initialized local parents.
+    """
+    return LocalSnapshot(
+        identifier=UUID(hex=identifier),
+        name=path,
+        author=author,
+        content_path=FilePath(content_paths[identifier]),
+        metadata=metadata.get(identifier, {}),
+        parents_remote=_get_remote_parents(identifier, parents),
+        parents_local=_get_local_parents(
+            identifier,
```

Comment:
Per the comment about the "parents" table (above), since we're *already* splitting these up anyway why not just store them in their own tables? Not that "efficiency"/optimization really matters here, but: 50% of the queries! ;)